src/transformers/processing_utils.py

```diff
@@ -1214,38 +1216,76 @@ def apply_chat_template(
-                for vision_info in visuals:
-                    if vision_info["type"] == "image":
-                        for key in ["image", "url", "path", "base64"]:
-                            if key in vision_info:
-                                images.append(load_image(vision_info[key]))
-                    elif vision_info["type"] == "video":
-                        for key in ["video", "url", "path"]:
-                            if key in vision_info:
-                                videos.append(
-                                    load_video(vision_info[key], num_frames=num_frames, backend=video_load_backend)
-                                )
+            batch_images, batch_videos = [], []
+            for conversation in conversations:
+                images, videos = [], []
+                for message in conversation:
+                    visuals = [content for content in message["content"] if content["type"] in ["image", "video"]]
+                    image_fnames = [
+                        vision_info[key]
+                        for vision_info in visuals
+                        for key in ["image", "url", "path", "base64"]
+                        if key in vision_info and vision_info["type"] == "image"
+                    ]
+                    video_fnames = [
+                        vision_info[key]
+                        for vision_info in visuals
+                        for key in ["video", "url", "path"]
+                        if key in vision_info and vision_info["type"] == "video"
+                    ]
+                    for fname in image_fnames:
+                        images.append(load_image(fname))
+                    for fname in video_fnames:
+                        if isinstance(fname, (list, tuple)) and isinstance(fname[0], str):
+                            video = [np.array(load_image(image_fname)).T for image_fname in fname]
+                            # create a 4D video because `load_video` always returns a 4D array
+                            video = np.stack(video)
+                        else:
+                            video = load_video(fname, num_frames=num_frames, fps=video_fps, backend=video_load_backend)
+                        videos.append(video)
+
+                # Currently all processors can accept accept nested list of batches, but not flat list of visuals
+                # So we'll make a batched list of images and let the processor handle it
+                if images:
+                    batch_images.append(images)
+                if videos:
+                    batch_videos.append(videos)
+
+            # Tokenizer's `apply_chat_template` never adds special tokens when tokenizing
+            # But processor's `apply_chat_template` didn't have an option to tokenize, so users had to format the prompt
+            # and pass it to the processor. Users thus never worried about special tokens relying on processor hadnling
```

Comment:
```suggestion
            # and pass it to the processor. Users thus never worried about special tokens relying on processor handling
```

Actually, [users had to be careful](https://huggingface.co/meta-llama/Llama-3.2-11B-Vision-Instruct/discussions/23#67a486fd77e167282b1f85c2) to use `add_special_tokens=False` when tokenizing the rendered template. Is this not the case for some VLMs?

Reply:
They had to but we never added that in the docs. For example in llava we didn't add `bos` in template but idefics copied llava, and added `bos` thus causing the problem of two special tokens

Reply:
I think the standard is for chat templates to handle special tokens, so idefics is doing it "right" and llava is an exception (that we have to provide BC for, of course). Perhaps we could default to the expected behaviour in our reference code, and treat exceptional cases based on some other property?

Reply:
yeah, I think we'll need to:

1. Update demo inference code when `tokenize=False`, and let users know why we pass `add_special_tokens=False`
2. Slowly see if updating templates in llava  doesn't increase comments from angry users. Might as well be bad idea, until we make `tokenize=True` a go-to default for everyone. Most users are stuck with old version of transformers
3. For any new model, make sure the `bos/eos` are in the template and not hacked from within processing code!